core/shared/src/main/scala/fs2/concurrent/Topic.scala

```diff
@@ -22,34 +22,39 @@
 package fs2
 package concurrent
 
-import cats.Eq
-import cats.effect.kernel.Concurrent
+import cats.effect._
 import cats.syntax.all._
+import cats.effect.std.{Queue => Q}
+import scala.collection.immutable.LongMap
 
-import fs2.internal.{SizedQueue, Unique}
-
-/** Asynchronous Topic.
-  *
-  * Topic allows you to distribute `A` published by arbitrary number of publishers to arbitrary number of subscribers.
+/** Topic allows you to distribute `A`s published by an arbitrary
+  * number of publishers to an arbitrary number of subscribers.
   *
-  * Topic has built-in back-pressure support implemented as maximum bound (`maxQueued`) that a subscriber is allowed to enqueue.
-  * Once that bound is hit, publishing may semantically block until the lagging subscriber consumes some of its queued elements.
+  * Topic has built-in back-pressure support implemented as a maximum
+  * bound (`maxQueued`) that a subscriber is allowed to enqueue.
   *
-  * Additionally the subscriber has possibility to terminate whenever size of enqueued elements is over certain size
-  * by using `subscribeSize`.
+  * Once that bound is hit, publishing may semantically block until
+  * the lagging subscriber consumes some of its queued elements.
   */
 abstract class Topic[F[_], A] { self =>
 
-  /** Publishes elements from source of `A` to this topic and emits a unit for each element published.
+  /** Publishes elements from source of `A` to this topic and emits a
+    * unit for each element published.
     * [[Pipe]] equivalent of `publish1`.
+    *
+    * TODO return `Nothing` for consistency?
     */
   def publish: Pipe[F, A, Unit]
 
   /** Publishes one `A` to topic.
     *
-    * This waits until `a` is published to all subscribers.
-    * If any of the subscribers is over the `maxQueued` limit, this will wait to complete until that subscriber processes
-    * enough of its elements such that `a` is enqueued.
+    * If any of the subscribers is over the `maxQueued` limit, this
+    * will wait to complete until that subscriber processes enough of
+    * its elements such that `a` is enqueued.
+    *
```

Comment:
```suggestion
    * This `F` operation deos not complete until after the 
    * given element has been enqued on all subscribers. 
    * If enqueuing on any subscriber is semantically blocked 
    * (because that subscriber is on its `maxQueued` limit), 
    * then this operation will also be blocked, waiting on that  
    * subscriber to dequeue an element.
```

One further clarification: does the whole `Topic` ensure that there is at most one call to `publish1` at a given point?

Reply:
no, you can have multiple producers on a Topic, although I imagine it to be rarer

Reply:
In that case, is there a risk that different subscribers may receive messages from different producers on a different order?

Reply:
Yeah, there is (and there's always been). I think fixing it would imply either enforcing single producer, if you still want full broadcast, or a KeyedTopic to do sharding a la Kafka

Reply:
@mpilquist wdyt about this? Should we enforce a single producer, or just document it and it's up to the user?